zndraw/server/tasks.py

```diff
@@ -338,6 +356,75 @@ def run_analysis(url: str, token: str, data: dict):
     vis.socket.emit("celery:task:emit", asdict(msg))
 
 
+def run_global_modifier(vis, NAME, data):
+    while True:
+        with Session(engine) as ses:
+            # get the available hosts for the modifier
+            modifier = (
+                ses.query(db_schema.GlobalModifier).filter_by(name=NAME).first()
+            )
+            host = (
+                ses.query(db_schema.GlobalModifierClient)
+                .filter_by(global_modifier=modifier, available=True)
+                .first()
+            )
+            assigned_hosts = ses.query(db_schema.GlobalModifierClient).filter_by(
+                global_modifier=modifier
+            ).count()
+        if assigned_hosts == 0:
+            msg = CeleryTaskData(
+            target=f"webclients_{vis.token}",
+            event="modifier:run:finished",
+            data=None,
+            )
+            vis.socket.emit("celery:task:emit", asdict(msg))
+            msg = CeleryTaskData(
+                target=f"webclients_{vis.token}",
+                event="message:alert",
+                data=f"Could not find any available modifier for {NAME}.",
+                disconnect=True,
+            )
+            vis.socket.emit("celery:task:emit", asdict(msg))
+            return
+        
+        if host is None:
+            vis.socket.sleep(1)
+            log.critical("No modifier available")
+            continue
+
+        # run the modifier
+        msg = CeleryTaskData(
+            target=host.sid,
+            event="modifier:run",
+            data={"params": data, "token": vis.token},
+        )
+        vis.socket.emit("celery:task:emit", asdict(msg))
+
+        # add additional 5 seconds for communication overhead
+        for _ in range(int(host.timeout + 5)):
+            if vis.socket.connected:
```

Comment:
We saw that `vis.socket.connected` is not really reliable. Ideally we would have a function in utils that takes in `vis` object and reliably returns `True` if it's connected `False` otherwise.